### ISSUE TITLE
Introduce leagues

### ISSUE THREAD

**Issue by mschmidt87** [jakobj/kicker_league#82]
At 1 point, we need to introduce the concept of leagues. @philipp-w and I discussed about this and had the following ideas:

1. Each player can belong to multiple leagues.
2. Each player belongs at least to the global league.
3. Each game is assigned to exactly one league.
4. We have to store a table for the relationship player <--> league in the database.
5. In the app, the user needs to select the league he wants to play in at the beginning and then gets a screen with the players of the league.

Please give comments.

**Comment by alperyeg** [comment]
I was also thinking about it and wanted to suggest. 
I think we need a new table 'leagues' for that, where you have player name or id and the league where the player is playing. 
In the app the player should choose his league at the start only if he has not chosen a league yet. After that it should be possible to switch from the menu. Also there should be a possibility to create a league.

**Comment by mschmidt87** [comment]
I think this table should be the other way around, like:
League --> Player

because players can play in multiple leagues.

**Comment by alperyeg** [comment]
Yes, also possible. But in the end if it is column or row in the table doesn't matter, I think. 
I saw that it is also possible to store  sets in a table, although the db operations are not efficient then.

**Comment by mschmidt87** [comment]
Yes, I actually though of this:
League1 Player1
League1 Player2
League2 Player1
...


But @philipp-w knows more about that, I guess.

**Comment by jakobj** [comment]
I would suggest a different approach: how about one table just listing all existent leagues and then per league a players and a games table. The app would query the one table with all leagues which would feed some drop down menu in the app. Once a player has selected a league, the data are fetched from the corresponding tables. If a user selects a league in which he is not yet in the players table, we offer him to create that entry with a "register" button. Before a player is visible in a league he needs to have participated in at least one game.

**Comment by mschmidt87** [comment]
Yes, sounds fine to me at first glance. This is also the easiest to implement, because it simply adds the league layer on top of the current structure.

**Comment by philipp-w** [comment]
Please not. This idea can lead to many problems. For example, by doing this, we would have to save all the players multiple time, which could lead to inconsistency. And further, we would create lots of tables, which would be very bad for the webinterface.

Before further speculating, please read:
http://de.wikipedia.org/wiki/Normalisierung_(Datenbank)

In my opinion, there is only one possible solution:
1 table for the leagues
1 table for the games
1 table for the players
1 table to assign players to leagues.

I'm totally willing to draw a UML diagram of that.

**Comment by jakobj** [comment]
you are completely right about the redundancy problem. if have not thought about the issues attached to duplicated player data. could you please create a diagram, because atm i don't see how it should work. especially i am concerned about saving all games in a single table. what is the advantage of that? does it not just make it harder to extract the relevant data for a single league? and why are many tables bad? and could you please reiterate your point about not storing multiple values in a single entry?

**Comment by philipp-w** [comment]
About multiple values in one entry:
Imagine we have tables like this:

leagues
id,        name
1,         global_league
2,         Institute 1
3,         Druegge Pitter

players
id,         name,         elo,        league_ids
1,          Peter         1100       {1, 3}                 (Peter plays in 2 different leagues)

Now, we want to extract all players from league 3 to calculate the ranking. How should that look like?

SELECT * FROM players WHERE league_ids = 3 (does not work)

I don't think there is a proper way to search in a list of values in one entry. But correct me if I'm wrong.

About many tables:
@jakobj idea was to have two tables per league. One for the players and one for the games. Imagine we have 1000 different leagues. The name of the tables are automatically chosen I assume. First, how to access them properly? I'm not entirely sure how to do that in a nice fashion.

Another simple reason against many tables: Image the web-interface with 1000 tables and you would like to chance something. I imagine the web-interface is then unusable.

What is the issue with all games in one table? Databases can grow quite a bit until the performance degrade:
http://stackoverflow.com/questions/1276/how-big-can-a-mysql-database-get-before-performance-starts-to-degrade

What problems do you see there?
